File: 54_spiral_matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> res;
        int row = matrix.size() - 1;
        int col = matrix[0].size() - 1;
        for(int i = 0, j = 0; i <= row && j <= col; i++, j++){
            //get top
            for(int k = j; k <= col; k++){
                res.push_back(matrix[i][k]);
            }
            if(i == row) return res; //when i == row there is only one row left 
            
            // get right
            for(int k = i + 1; k <= row; k++){
                res.push_back(matrix[k][col]);
            }
            if(j == col) return res; //when j == col there is only one col left
            
            // get bottom
            for(int k = col - 1; k >= j; k--){
                res.push_back(matrix[row][k]);
            }
            
            //get left
            for(int k = row - 1; k > i; k--){
                res.push_back(matrix[k][i]);
            }
            row--, col--;
        }
        
        return res;
    }
};

File: 54_spiral_matrix.py
def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        if not matrix:
            return []
        R = len(matrix)
        C = len(matrix[0])
        x = 0
        ans = []
        
        visited = [False] * (R*C)
        while x < min(R, C):
            for i in range(x, C-x):
                if not visited[x*C + i]:    
                    ans.append(matrix[x][i])
                    visited[x*C + i] = True
                    
            
            for j in range(x+1,R-x):
                if not visited[j*C + C-x-1]:
                    ans.append(matrix[j][C-x-1])
                    visited[j*C + C-x-1] = True
            
            for k in range(C-x-2, x-1, -1):
                if not visited[(R-x-1)*C + k]:
                    ans.append(matrix[R-x-1][k])
                    visited[(R-x-1)*C + k] = True
            
            for l in range(R-x-2, x, -1):
                if not visited[l*C + x]:
                    ans.append(matrix[l][x])
                    visited[l*C + x] = True
                    
            
            x += 1
                
        return ans